### PR TITLE
Added AssetCap for all the assets

### PR DESCRIPTION
**Pull request by adi44**
Comptroller Storage : Added a New variable of Address Datatype known as AssetCapGuardian
Added A new mapping for assets => assetCap

Comptroller : Added a functionality that allows admin or AssetCapGuardian to Put the assetCap for every asset in Fuse Contract. 
Added a functionality that sets the AssetCapGaurdian

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -927,6 +939,44 @@ contract Comptroller is ComptrollerV2Storage, ComptrollerInterface, ComptrollerE
...
 
+    /**
+      * @notice Set the given Asset caps for the given cToken markets. AssetCapSupply that brings total supply to or above asset cap will revert.
+      * @dev Admin or AssetCapGuardian function to set the Asset caps. A Asset cap of 0 corresponds to unlimited supply.
+      * @param cTokens The addresses of the markets (tokens) to change the supply caps for
+      * @param newAssetCaps The new borrow cap values in underlying to be set. A value of 0 corresponds to unlimited supply.
+      */
+
+        function _setMarketAssetCaps(CToken[] calldata cTokens, uint[] calldata newAssetCaps) external {
+    	require(msg.sender == admin || msg.sender == AssetCapGuardian, "only admin or borrow cap guardian can set borrow caps");
```
Comment (transmissions11) [inline contracts/Comptroller.sol]: The use of require() does not conform to Compound's error handling system

File: contracts/Comptroller.sol [hunk 2]
```diff
@@ -927,6 +939,44 @@ contract Comptroller is ComptrollerV2Storage, ComptrollerInterface, ComptrollerE
...
+            emit NewAssetCap(cTokens[i], newAssetCaps[i]);
+        }
+    }
+
+    /**
+     * @notice Admin function to change the Asset Cap Guardian
+     * @param newAssetCapGuardian The address of the new Asset Cap Guardian
+     */
+    function _setAssetCapGuardian(address newAssetCapGuardian) external {
+        require(msg.sender == admin, "only admin can set asset cap guardian");
```
Comment (transmissions11) [inline contracts/Comptroller.sol]: The use of require() does not conform to Compound's error handling system

File: contracts/Comptroller.sol [hunk 3]
```diff
@@ -927,6 +939,44 @@ contract Comptroller is ComptrollerV2Storage, ComptrollerInterface, ComptrollerE
...
+
+        // Save current value for inclusion in log
+        address oldAssetCapGuardian = AssetCapGuardian;
```
Comment (transmissions11) [inline contracts/Comptroller.sol:984]: why not just emit the event before you make the change? that will save you an MLOAD and MSTORE.

**Review by transmissions11 — CHANGES_REQUESTED** [review]

File: contracts/ComptrollerStorage.sol [hunk 4]
```diff
@@ -137,13 +137,18 @@ contract ComptrollerV2Storage is ComptrollerV1Storage {
     /// @notice Indexes of account addresses in the `whitelistArray` array
     mapping(address => uint256) internal whitelistIndexes;
 
+    /// @notice Maps CTokens to Max Supply cap indicating what should be the maximum supply cap for all assets
```
Comment (transmissions11) [inline contracts/ComptrollerStorage.sol]: This comment is misleading. It maps each cToken to **it's own supply cap**, not "the supply cap for all assets"

File: contracts/Comptroller.sol [hunk 5]
```diff
@@ -927,6 +939,44 @@ contract Comptroller is ComptrollerV2Storage, ComptrollerInterface, ComptrollerE
...
+      * @param newAssetCaps The new borrow cap values in underlying to be set. A value of 0 corresponds to unlimited supply.
+      */
+
+        function _setMarketAssetCaps(CToken[] calldata cTokens, uint[] calldata newAssetCaps) external {
+    	require(msg.sender == admin || msg.sender == AssetCapGuardian, "only admin or borrow cap guardian can set borrow caps"); 
+
+        uint numMarkets = cTokens.length;
+        uint numAssetCaps = newAssetCaps.length;
+
+        require(numMarkets != 0 && numMarkets == numAssetCaps, "invalid input");
```
Comment (transmissions11) [inline contracts/Comptroller.sol]: The use of require() does not conform to Compound's error handling system

File: contracts/Comptroller.sol (hunk 3 above)
Comment (adi44) [reply contracts/Comptroller.sol:984]: Ohkay, I will make required changes

File: contracts/ComptrollerStorage.sol (hunk 4 above)
Comment (adi44) [reply contracts/ComptrollerStorage.sol]: Yeah! I agree I will make changes

File: contracts/Comptroller.sol (hunk 1 above)
Comment (adi44) [reply contracts/Comptroller.sol]: ohkay! I will use if statement and error handling by Compound's error handling system

Comment (transmissions11) [reply contracts/Comptroller.sol]: thanks!

File: contracts/Comptroller.sol (hunk 3 above)
Comment (transmissions11) [reply contracts/Comptroller.sol:984]: sounds good!

File: contracts/ComptrollerStorage.sol (hunk 4 above)
Comment (transmissions11) [reply contracts/ComptrollerStorage.sol]: great!

File: contracts/Comptroller.sol (hunk 3 above)
Comment (adi44) [reply contracts/Comptroller.sol:984]: made Changes @TransmissionsDev

File: contracts/Comptroller.sol (hunk 2 above)
Comment (adi44) [reply contracts/Comptroller.sol]: made Changes @TransmissionsDev

File: contracts/Comptroller.sol (hunk 1 above)
Comment (adi44) [reply contracts/Comptroller.sol]: made Changes @TransmissionsDev

File: contracts/Comptroller.sol (hunk 5 above)
Comment (adi44) [reply contracts/Comptroller.sol]: made Changes @TransmissionsDev

File: contracts/ComptrollerStorage.sol (hunk 4 above)
Comment (adi44) [reply contracts/ComptrollerStorage.sol]: made Changes @TransmissionsDev

**Pull request closed without merging**